src/ucp/core/ucp_rkey.h

```diff
@@ -84,28 +96,39 @@ typedef struct {
  * The array itself contains only the MDs specified in md_map, without gaps.
  */
 typedef struct ucp_rkey {
-    /* cached values for the most recent endpoint configuration */
-    struct {
-        ucp_worker_cfg_index_t    ep_cfg_index; /* EP configuration relevant for the cache */
-        ucp_lane_index_t          rma_lane;     /* Lane to use for RMAs */
-        ucp_lane_index_t          amo_lane;     /* Lane to use for AMOs */
-        ssize_t                   max_put_short;/* Cached value of max_put_short */
-        uct_rkey_t                rma_rkey;     /* Key to use for RMAs */
-        uct_rkey_t                amo_rkey;     /* Key to use for AMOs */
-        ucp_amo_proto_t           *amo_proto;   /* Protocol for AMOs */
-        ucp_rma_proto_t           *rma_proto;   /* Protocol for RMAs */
-    } cache;
-    ucp_md_map_t                  md_map;       /* Which *remote* MDs have valid memory handles */
-    ucs_memory_type_t             mem_type;     /* Memory type of remote key memory */
-    uint8_t                       flags;        /* Rkey flags */
-    ucp_worker_cfg_index_t        cfg_index;    /* Rkey configuration index */
+    union {
+        /* cached values for the most recent endpoint configuration */
+        struct {
+            uint8_t                   flags;          /* Rkey flags */
```

Comment:
do we need "Rkey flags" for both cache part and common part (see 113 line) of `ucp_rkey_t`?

Reply:
![image](https://user-images.githubusercontent.com/17932065/128703342-24b4ace9-f03f-44b6-ab3f-92ee67a2c097.png)
`Flags` use in both cases, and logically it can be placed out of the union. But since the union has gaps we may place it here without affecting the general `ucp_rkey_t` size. Also, it is not affected the code base, because in both cases we can be accessed flags by `rkey->flags`.
I think makes sense to add assert or some checking if (rkey->flags == rkey->cache.flags) have different addresses. It guarantees that nothing crashed if rkey_t will be modified in the future.

Reply:
👍 thanks for the explanation.
wdyt about checking `ucs_assert((&rkey->flags == &rkey->cache.flags) && (ucs_field_sizeof(ucp_rkey_t, flags) == ucs_field_sizeof(ucp_rkey_t, cache.flags)))` (i.e. checking that addresses and lengths of the fields are same) in the `ucp_rkey_resolve_inner()` instead?

Reply:
thanks, will be added.

Reply:
please use UCS_STATIC_ASSERT

Reply:
> please use UCS_STATIC_ASSERT

to use `UCS_STATIC_ASSERT`, it should be:
```
UCS_STATIC_ASSERT((ucs_offsetof(rkey, flags) == ucs_offsetof(rkey, cache.flags)) && (ucs_field_sizeof(ucp_rkey_t, flags) == ucs_field_sizeof(ucp_rkey_t, cache.flags)))
```